File: 1918-maximum-score-of-a-good-subarray/maximum-score-of-a-good-subarray.js
/**
 * @param {number[]} nums
 * @param {number} k
 * @return {number}
 */
var maximumScore = function(nums, k) {

    n = nums.length
    prevSmall = Array(n).fill(0)
    nextSmall = Array(n).fill(0)
    stk =[[-1,-1]]
    for(let i = 0 ;i<nums.length ;i++){

        num  = nums[i]
        while(stk.length>0  && stk[stk.length-1][1]>= num){


            stk.pop()
        }
        
        prevSmall[i] = stk[stk.length-1][0]
        stk.push([i,num])

    }
        
    stk =[[n,-1]]
    for(let i= nums.length-1;i>=0 ; i--){

        num  = nums[i]
        
        while(stk.length>0 &&  stk[stk.length-1][1]>= num){
            stk.pop()
        

        }
            
        nextSmall[i] =  stk[stk.length-1][0]
        stk.push([ i,num])
    }
        
    let ans =  nums[k]
    for(let p = 0; p<nums.length;p++){
        num  = nums[ p]
        i = prevSmall[p]+1
        j = nextSmall[p]-1
        //if the subarray is good subarray , take the score into accounte 
        if (i<=k && k<=j){
            ans =Math.max(ans, num *(j-i+1))

        }
    }
            
    return ans
    
};

File: 1918-maximum-score-of-a-good-subarray/maximum-score-of-a-good-subarray.java
class Solution {
    public int maximumScore(int[] nums, int k) {


        //  '''

        //     find the index of prevsmall and next small number  for each number 

        //     then iterate on the nums array , and each num as minimum of the  subarray and score  and also including k 
        //     find the score and store max score

        // '''


    int n = nums.length;
    int[] prevSmall = new int[n];
    int [] nextSmall = new int[n];
    
    Stack<int[]> stk = new Stack<int[]>();
    stk.push(new int[]{-1,-1});
        for(int  i = 0 ;i<nums.length ;i++){

            int num  = nums[i];
            while(!stk.empty() && stk.peek()[1]>= num){


                stk.pop();
            }
            
            prevSmall[i] = stk.peek()[0];
            stk.push(new int[]{i,num});

        }
            
        stk.clear();
        stk.push(new int[]{n,-1});
        for(int i= nums.length-1;i>=0 ; i--){

            int num  = nums[i];
            
            while(!stk.empty()  && stk.peek()[1]>= num){
                stk.pop();
            

            }
              
            nextSmall[i] = stk.peek()[0];
            stk.push(new int[]{i,num});
        }
            
        int  ans =  nums[k];
        for(int p = 0; p<nums.length;p++){
            int num  = nums[ p];
            int i = prevSmall[p]+1;
            int j = nextSmall[p]-1;
            //if the subarray is good subarray , take the score into accounte 
            if (i<=k && k<=j){
                ans =Math.max(ans, num *(j-i+1));

            }
        }
               
        return ans;
    

}
}